File: config_editor.js
const fs = require("fs");
const readline = require("readline");
const path = require("path");

const DEFAULT_CONFIG_FILE = "config.json";

function parseArgs() {
    const args = process.argv.slice(2);
    let configPath = DEFAULT_CONFIG_FILE;

    for (let i = 0; i < args.length; i++) {
        if ((args[i] === "-c" || args[i] === "--config") && i + 1 < args.length) {
            configPath = args[i + 1];
            break;
        }
    }

    return { configPath };
}

function loadConfig(configPath) {
    try {
        const configData = fs.readFileSync(configPath, "utf8");
        return JSON.parse(configData);
    } catch (error) {
        console.error(`Error loading configuration: ${error.message}`);
        return {
            proxy_file: "proxy.txt",
            output_file: "working_proxies.txt",
            test_url: "http://www.google.com",
            timeout: 5,
            concurrent_checks: 20,
            save_to_input_file: false,
        };
    }
}

function saveConfig(config, configPath) {
    try {
        fs.writeFileSync(configPath, JSON.stringify(config, null, 4));
        console.log(`Configuration saved to file: ${configPath}`);
        return true;
    } catch (error) {
        console.error(`Error saving configuration: ${error.message}`);
        return false;
    }
}

function displayConfig(config) {
    console.log("\nCurrent configuration:");
    console.log("-".repeat(30));
    for (const [key, value] of Object.entries(config)) {
        console.log(`${key}: ${value}`);
    }
    console.log("-".repeat(30));
}

function createInterface() {
    return readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });
}

function question(rl, query) {
    return new Promise((resolve) => {
        rl.question(query, (answer) => {
            resolve(answer);
        });
    });
}

async function editConfig(configPath) {
    const config = loadConfig(configPath);
    displayConfig(config);

    const rl = createInterface();

    console.log("\nEditing configuration:");
    console.log("Leave the field empty to keep the current value");

    const proxyFile = await question(rl, `Proxy file [${config.proxy_file}]: `);
    if (proxyFile) config.proxy_file = proxyFile;

    const outputFile = await question(rl, `Output file [${config.output_file}]: `);
    if (outputFile) config.output_file = outputFile;

    const testUrl = await question(rl, `Test URL [${config.test_url}]: `);
    if (testUrl) config.test_url = testUrl;

    const timeout = await question(rl, `Timeout in seconds [${config.timeout}]: `);
    if (timeout) {
        const timeoutNum = parseInt(timeout);
        if (!isNaN(timeoutNum)) {
            config.timeout = timeoutNum;
        } else {
            console.log("Invalid timeout format. Using the previous value.");
        }
    }

    const concurrentChecks = await question(rl, `Number of concurrent checks [${config.concurrent_checks}]: `);
    if (concurrentChecks) {
        const concurrentChecksNum = parseInt(concurrentChecks);
        if (!isNaN(concurrentChecksNum)) {
            config.concurrent_checks = concurrentChecksNum;
        } else {
            console.log("Invalid number format. Using the previous value.");
        }
    }

    const saveToInputFile = await question(
        rl,
        `Save working proxies to input file (y/n) [${config.save_to_input_file ? "y" : "n"}]: `
    );
    if (
        saveToInputFile.toLowerCase() === "y" ||
        saveToInputFile.toLowerCase() === "yes"
    ) {
        config.save_to_input_file = true;
    } else if (
        saveToInputFile.toLowerCase() === "n" ||
        saveToInputFile.toLowerCase() === "no"
    ) {
        config.save_to_input_file = false;
    }

    rl.close();

    if (saveConfig(config, configPath)) {
        console.log("Configuration successfully updated!");
        displayConfig(config);
    }
}

async function main() {
    const { configPath } = parseArgs();

    if (!fs.existsSync(configPath)) {
        const defaultConfig = loadConfig(configPath);
        saveConfig(defaultConfig, configPath);
        console.log(`Created default configuration file: ${configPath}`);
    }

    await editConfig(configPath);
}

main().catch((error) => {
    console.error("Error executing the program:", error);
});

File: config_editor.py
import json
import os
import argparse

DEFAULT_CONFIG_FILE = "config.json"


def parse_args():
    parser = argparse.ArgumentParser(
        description="Configuration editor for Proxy Checker"
    )
    parser.add_argument(
        "-c",
        "--config",
        default=DEFAULT_CONFIG_FILE,
        help=f"Path to the configuration file (default: {DEFAULT_CONFIG_FILE})",
    )
    return parser.parse_args()


def load_config(config_file):
    try:
        with open(config_file, "r") as file:
            return json.load(file)
    except Exception as e:
        print(f"Error loading config file: {e}")
        return {
            "proxy_url": "https://raw.githubusercontent.com/monosans/proxy-list/refs/heads/main/proxies/all.txt",
            "proxy_file": "proxy.txt",
            "output_file": "working_proxies.txt",
            "test_url": "http://www.google.com",
            "timeout": 5,
            "concurrent_checks": 20,
            "save_to_input_file": False,
        }


def save_config(config_data, config_file):
    try:
        with open(config_file, "w") as file:
            json.dump(config_data, file, indent=4)
        print(f"Configuration saved to {config_file}")
        return True
    except Exception as e:
        print(f"Error saving config file: {e}")
        return False


def display_config(config):
    print("\nCurrent configuration:")
    print("-" * 30)
    for key, value in config.items():
        print(f"{key}: {value}")
    print("-" * 30)


def edit_config(config_file):
    config = load_config(config_file)
    display_config(config)

    print("\nEditing configuration:")
    print("Leave the field empty to keep the current value")

    config["proxy_url"] = input(
        f"Proxy URL [{config.get('proxy_url', '')}]: "
    ) or config.get("proxy_url", "")
    config["proxy_file"] = (
        input(f"Proxy file [{config['proxy_file']}]: ") or config["proxy_file"]
    )
    config["output_file"] = (
        input(f"Output file [{config['output_file']}]: ") or config["output_file"]
    )
    config["test_url"] = (
        input(f"Test URL [{config['test_url']}]: ") or config["test_url"]
    )

    try:
        timeout_input = input(f"Timeout in seconds [{config['timeout']}]: ")
        if timeout_input:
            config["timeout"] = int(timeout_input)
    except ValueError:
        print("Invalid timeout format. Using the previous value.")

    try:
        concurrent_input = input(
            f"Number of concurrent checks [{config['concurrent_checks']}]: "
        )
        if concurrent_input:
            config["concurrent_checks"] = int(concurrent_input)
    except ValueError:
        print("Invalid number format. Using the previous value.")

    save_to_input_file = config.get("save_to_input_file", False)
    save_input = input(
        f"Save working proxies to input file (y/n) [{('y' if save_to_input_file else 'n')}]: "
    )
    if save_input.lower() in ["y", "yes"]:
        config["save_to_input_file"] = True
    elif save_input.lower() in ["n", "no"]:
        config["save_to_input_file"] = False

    if save_config(config, config_file):
        print("Configuration successfully updated!")
        display_config(config)


if __name__ == "__main__":
    args = parse_args()
    config_file = args.config

    if not os.path.exists(config_file):
        default_config = load_config(config_file)
        save_config(default_config, config_file)
        print(f"Created default configuration file: {config_file}")

    edit_config(config_file)
